GitVersionCore/GitVersionContext.cs

```diff
@@ -9,71 +9,73 @@
         {
             Repository = repository;
-            Configuration = configuration;
-            IsContextForTrackedBranchesOnly = isForTrackingBranchOnly;
+            this.configuration = configuration;
+            OnlyEvaluateTrackedBranches = onlyEvaluateTrackedBranches;
 
             if (currentBranch == null)
-                return;
+                throw new InvalidOperationException("Need a branch to operate on");
 
             CurrentCommit = currentBranch.Tip;
-
-            if (repository != null && currentBranch.IsDetachedHead())
+            if (currentBranch.IsDetachedHead())
             {
-                CurrentBranch = GetBranchesContainingCommit(CurrentCommit.Sha).OnlyOrDefault() ?? currentBranch;
+                CurrentBranch = CurrentCommit.GetBranchesContainingCommit(repository, OnlyEvaluateTrackedBranches).OnlyOrDefault() ?? currentBranch;
             }
             else
             {
                 CurrentBranch = currentBranch;
             }
+
+            CalculateEffectiveConfiguration();
+
+            CurrentCommitTaggedVersion = repository.Tags
+                .SelectMany(t =>
+                {
+                    SemanticVersion version;
+                    if (t.PeeledTarget() == CurrentCommit && SemanticVersion.TryParse(t.Name, Configuration.GitTagPrefix, out version))
+                        return new[] { version };
+                    return new SemanticVersion[0];
+                })
+                .Max();
+            IsCurrentCommitTagged = CurrentCommitTaggedVersion != null;
         }
 
-        public Config Configuration { get; private set; }
+        public SemanticVersion CurrentCommitTaggedVersion { get; private set; }
+        public bool OnlyEvaluateTrackedBranches { get; private set; }
+        public EffectiveConfiguration Configuration { get; private set; }
         public IRepository Repository { get; private set; }
         public Branch CurrentBranch { get; private set; }
         public Commit CurrentCommit { get; private set; }
+        public bool IsCurrentCommitTagged { get; private set; }
 
-        readonly bool IsContextForTrackedBranchesOnly = true;
-
-
-        IEnumerable<Branch> GetBranchesContainingCommit(string commitSha)
+        void CalculateEffectiveConfiguration()
         {
-            var directBranchHasBeenFound = false;
-            foreach (var branch in Repository.Branches)
-            {
-                if (branch.Tip.Sha != commitSha || (IsContextForTrackedBranchesOnly && !branch.IsTracking))
-                {
-                    continue;
-                }
+            var currentBranchConfig = BranchConfigurationCalculator.GetBranchConfiguration(CurrentCommit, Repository, OnlyEvaluateTrackedBranches, configuration, CurrentBranch);
 
-                directBranchHasBeenFound = true;
-                yield return branch;
-            }
+            // Versioning mode drills down, if top level is specified then it takes priority
```

Comment:
If versioning mode is specified at the top level, it's not possible to override per branch?